src/app/workflow/launch-third-party/launch-third-party.component.ts

```diff
@@ -321,4 +344,48 @@ export class LaunchThirdPartyComponent extends Base implements OnChanges, OnInit
     }
     return `Export this workflow version to ${platform}.`;
   }
+
+  launchToCodespace() {
+    const correctDevcontainerPath = this.computeCorrectDevcontainerPath(this.devcontainers);
+    if (correctDevcontainerPath) {
+      this.openNewCodespaceWindow(correctDevcontainerPath);
+    } else {
+      this.displayLaunchToCodespaceDialog(this.devcontainers.length > 0)
+        .afterClosed()
+        .subscribe((shouldLaunch: boolean) => {
+          if (shouldLaunch) {
+            this.openNewCodespaceWindow(undefined);
+          }
+        });
+    }
+  }
+
+  private displayLaunchToCodespaceDialog(hasDevcontainer: boolean) {
+    return this.dialog.open(LaunchToCodespaceDialogComponent, {
+      width: bootstrap4largeModalSize,
+      data: { entry: this.workflow, hasDevcontainer: hasDevcontainer },
+    });
+  }
+
+  private computeCorrectDevcontainerPath(devcontainers: SourceFile[]): string | undefined {
+    // Return the path, relative to root, of the devcontainer with a body that contains the absolute notebook path as a substring,
+    // or undefined if there is no such devcontainer
+    return devcontainers.find((file) => file.content.includes(this.selectedVersion.workflow_path))?.absolutePath?.substring(1);
+  }
+
+  private openNewCodespaceWindow(devcontainerPath: string | undefined) {
+    let url: string =
+      this.config.GITHUB_CODESPACES_IMPORT_URL +
+      '?hide_repo_select=true' +
+      '&ref=' +
+      this.selectedVersionNameAsQueryValue +
```

Comment:
Does this need to be encoded? Looks like GitHub branch names are fairly restrictive, so probably not, but making sure.

Reply:
Already is encoded, all the ...`asQueryValue` variables are, see https://github.com/dockstore/dockstore-ui2/pull/1865/files#diff-4b05662ab7343f5beca9de782e5618b59a82a5c0bbde36471800e2c64daee254R298-R303